File: IsSubsequence.cpp
#include <iostream>
#include <string>
using namespace std;

bool isSubsequence(string s, string t) {
    int i = 0, j = 0;
    while (i < s.size() && j < t.size()) {
        if (s[i] == t[j]) {
            i++; // Move in s only if characters match
        }
        j++; // Always move in t
    }
    return i == s.size(); // All characters in s matched
}

int main() {
    string s = "abc";
    string t = "ahbgdc";

    if (isSubsequence(s, t)) cout << "s is a subsequence of t\n";
    else cout << "s is NOT a subsequence of t\n";

    return 0;
}

File: Is Subsequence.java
public class Solution {
    public boolean isSubsequence(String s, String t) {
        int i = 0; // pointer for s
        int j = 0; // pointer for t

        // Traverse both strings
        while (i < s.length() && j < t.length()) {
            if (s.charAt(i) == t.charAt(j)) {
                i++; // move to next char in s if match found
            }
            j++; // always move in t
        }

        // if we reached the end of s, all chars matched
        return i == s.length();
    }

    public static void main(String[] args) {
        Solution sol = new Solution();

        System.out.println(sol.isSubsequence("abc", "ahbgdc")); // true
        System.out.println(sol.isSubsequence("axc", "ahbgdc")); // false
    }
}
